File: 2021/day_10/day.rb

def calc(data)
    opens = '[({<'
    matches = {
        '['=> ']',
        '('=> ')',
        '{'=> '}',
        '<'=> '>',
        ']'=> '[',
        ')'=> '(',
        '}'=> '{',
        '>'=> '<',
    }
    points = {
        ')'=> 3,
        ']'=> 57,
        '}'=> 1197,
        '>'=> 25137,
    }
    point_count = 0
    data.each { |d|
      stack = []
      d.strip.each_char { |c|
        if opens.include?(c)
          stack.append(c)
        else
          if stack[-1] == matches[c]
            stack.pop
          else
            point_count += points[c]
            break
          end
        end
      }
    }
    point_count
end

def calc2(data)
    opens = '[({<'
    matches = {
        '['=> ']',
        '('=> ')',
        '{'=> '}',
        '<'=> '>',
        ']'=> '[',
        ')'=> '(',
        '}'=> '{',
        '>'=> '<',
    }
    points = {
        ')'=> 1,
        ']'=> 2,
        '}'=> 3,
        '>'=> 4,
    }
    scores = []
    data.each { |d|
      stack = []
      d.strip.each_char { |c|
        if opens.include?(c)
          stack.append(c)
        else
          if stack[-1] == matches[c]
            stack.pop
          else
            stack = []
            break
          end
        end
      }
      if stack.empty?
        next
      end
      point_count = 0
      stack.reverse_each { |s|
        point_count *= 5
        point_count += points[matches[s]]
      }
      scores.append(point_count)
    }
    scores.sort[scores.size/2]
end


describe "Advent of code tests pass" do
    let(:sample) {[
"[({(<(())[]>[[{[]{<()<>>",
"[(()[<>])]({[<{<<[]>>(",
"{([(<{}[<>[]}>{[]{[(<()>",
"(((({<>}<{<{<>}{[]{[]{}",
"[[<[([]))<([[{}[[()]]]",
"[{[{({}]{}}([{[{{{}}([]",
"{<[[]]>}<{[{[{[]{()[[[]",
"[<(<(<(<{}))><([]([]()",
"<{([([[(<>()){}]>(<<{{",
"<{([{{}}[<[[[<>{}]]]>[]]",
]}

  context "first" do
    it "Processes sample properly" do
      expect(calc sample).to eq(26397)
    end

    it "Processes file" do
      expect(calc File.open("data").readlines).to eq(318099)
    end
  end

  context "second" do
    it "Processes sample properly" do
      expect(calc2 sample).to eq(288957)
    end

    it "Processes file" do
      expect(calc2 File.open("data").readlines).to eq(2389738699)
    end
  end
end

File: 2021/day_10/day.py
import unittest
import re
from collections import defaultdict
from collections import Counter
from numpy import sign

def calc(data):
    opens = '[({<'
    matches = {
        '[': ']',
        '(': ')',
        '{': '}',
        '<': '>',
        ']': '[',
        ')': '(',
        '}': '{',
        '>': '<',
    }
    points = {
        ')': 3,
        ']': 57,
        '}': 1197,
        '>': 25137,
    }
    point_count = 0
    for d in data:
        #print d
        stack = []
        for c in d.strip():
            #print stack
            if c in opens:
                stack.append(c)
            else:
                if stack[-1] == matches[c]:
                    stack.pop()
                else:
                    point_count += points[c]
                    #print "%r does not match %r" % (c, stack[0])
                    break
            
    return point_count

def calc2(data):
    opens = '[({<'
    matches = {
        '[': ']',
        '(': ')',
        '{': '}',
        '<': '>',
        ']': '[',
        ')': '(',
        '}': '{',
        '>': '<',
    }
    points = {
        ')': 1,
        ']': 2,
        '}': 3,
        '>': 4,
    }
    scores = []
    for d in data:
        #print d
        stack = []
        for c in d.strip():
            #print stack
            if c in opens:
                stack.append(c)
            else:
                if stack[-1] == matches[c]:
                    stack.pop()
                else:
                    #print "%r does not match %r" % (c, stack[0])
                    print 
                    stack = []
                    break
        if len(stack) == 0:
            print "Discarding %r %r" % (stack, d)
            continue
        print "Calculating score"
        print stack
        point_count = 0
        for s in stack[::-1]:
            point_count *= 5
            point_count += points[matches[s]]
        print point_count
        print scores
        scores.append(point_count)
    print "Scores %r" % scores
    scores.sort()
    return scores[len(scores)/2]

class TestDay(unittest.TestCase):
    sample = [
"[({(<(())[]>[[{[]{<()<>>",
"[(()[<>])]({[<{<<[]>>(",
"{([(<{}[<>[]}>{[]{[(<()>",
"(((({<>}<{<{<>}{[]{[]{}",
"[[<[([]))<([[{}[[()]]]",
"[{[{({}]{}}([{[{{{}}([]",
"{<[[]]>}<{[{[{[]{()[[[]",
"[<(<(<(<{}))><([]([]()",
"<{([([[(<>()){}]>(<<{{",
"<{([{{}}[<[[[<>{}]]]>[]]",
]

    def test_first_sample_data(self):
        self.assertEqual(calc(self.sample), 26397)

    def test_first_final_data(self):
        self.assertEqual(calc(open('data').readlines()), 318099)

    def test_second_challenge_sample(self):
        self.assertEqual(calc2(self.sample), 288957)

    def test_second_final_data_2(self):
        self.assertEqual(calc2(open('data').readlines()), 2389738699)

if __name__ == '__main__':
    unittest.main()
